concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limiter/LifoBlockingLimiter.java

```diff
@@ -0,0 +1,214 @@
+            return new LifoBlockingLimiter<ContextT>(this);
+        }
+    }
+    
+    public static <ContextT> Builder<ContextT> newBuilder(Limiter<ContextT> delegate) {
+        return new Builder<ContextT>(delegate);
+    }
+    
+    private final Limiter<ContextT> delegate;
+    
+    private static class Event<ContextT> {
+        private volatile Optional<Listener> listener;
+        private final CountDownLatch latch = new CountDownLatch(1);
+        private ContextT context;
+        
+        public Event(ContextT context) {
+            this.context = context;
+        }
+
+        public boolean await(long timeout, TimeUnit unit) throws InterruptedException {
+            return latch.await(timeout, unit);
+        }
+        
+        public void set(Optional<Listener> listener) {
+            this.listener = listener;
+            latch.countDown();
+        }
+        
+    }
+    
+    /**
+     * Lock used to block and unblock callers as the limit is reached
+     */
+    private final Deque<Event<ContextT>> backlog = new LinkedList<>();
+    
+    private final AtomicInteger backlogCounter = new AtomicInteger();
+    
+    private final int backlogSize;
+    
+    private final Function<ContextT, Long> backlogTimeoutMillis;
+    
+    private final Object lock = new Object();
+    
+    private LifoBlockingLimiter(Builder<ContextT> builder) {
+        this.delegate = builder.delegate;
+        this.backlogSize = builder.maxBacklogSize;
+        backlogTimeoutMillis = builder.maxBacklogTimeoutMillis;
+    }
+    
+    private Optional<Listener> tryAcquire(ContextT context) {
+        // Try to acquire a token and return immediately if successful
+        final Optional<Listener> listener = delegate.acquire(context);
+        if (listener.isPresent()) {
+            return listener;
+        }
+
+        // Otherwise block until the limit has been released or we timeout 
+        if (backlogCounter.get() > this.backlogSize) {
+            return Optional.empty();
+        }
+        backlogCounter.incrementAndGet();
+        final Event<ContextT> event = new Event<>(context);
+        
+        try {
+            synchronized (lock) {
+                backlog.addFirst(event);
+            }
+            
+            if (!event.await(backlogTimeoutMillis.apply(context), TimeUnit.MILLISECONDS)) {
+                synchronized (lock) {
+                    backlog.removeFirstOccurrence(event);
+                }
+                return Optional.empty();
+            }
+            return event.listener;
+        } catch (InterruptedException e) {
+            synchronized (lock) {
+                backlog.removeFirstOccurrence(event);
+            }
+            Thread.currentThread().interrupt();
+            return Optional.empty();
+        } finally {
+            backlogCounter.decrementAndGet();
+        }
+    }
+    
+    private void unblock() {
+        synchronized (lock) {
```

Comment:
every thread will need to acquire this lock whether they were interacting with the backlog previous right? Does this happen after we've replied to the customer, or could contention here lead to delay in already successful responses?

Reply:
Happens after we reply to the consumer